File: node_server/masterServer.js
const net = require('net');
const {peekFrameSize, parseFrame, buildOpResponse, buildPingResponse, buildEvent, ParamKey, EgMsgType, OpCode, photonByte} = require('./protocol');

let sessionIdCounter = 1;

class PlayerSession {
  constructor(socket) {
    this.socket = socket;
    this.buffer = Buffer.alloc(0);
    this.id = sessionIdCounter++;
    this.isAuthenticated = false;
    this.userId = '';
    this.nickname = '';
    this.connectTime = Date.now();
    this.socket.on('error', err => {
      // suppress per-socket errors in production
    });
  }

  appendAndExtract(data) {
    this.buffer = Buffer.concat([this.buffer, data]);
    const frames = [];
    while (true) {
      const sz = peekFrameSize(this.buffer);
      if (sz < 0) break;
      frames.push(this.buffer.slice(0, sz));
      this.buffer = this.buffer.slice(sz);
    }
    return frames;
  }

  send(buf) {
    this.socket.write(buf);
  }
}

class RoomInfo {
  constructor(name) {
    this.name = name;
    this.maxPlayers = 10;          // matches C# default
    this.isOpen = true;
    this.isVisible = true;
    this.map = 'FreeRun3_1';      // C# default map
    this.version = '';
    this.mode = '0';              // C# default mode (TDM)
    this.players = new Set();
    this._actorCounter = 0;
  }
  assignActorNr() { return ++this._actorCounter; }
  get playerCount() { return this.players.size; }
  getGamePropsHashtable() {
    // Keys that the client reads as (byte) MUST be serialized as GpType.Byte.
    // Unboxing int→byte in C# throws InvalidCastException, crashing room setup.
    return {
      255: photonByte(this.maxPlayers),   // (byte)MaxPlayers
      253: this.isOpen,                   // bool IsOpen
      254: this.isVisible,                // bool IsVisible
      252: photonByte(this.playerCount),  // (byte)PlayerCount
      251: false,                         // removedFromList
      map:     this.map,
      version: this.version,
      mode:    this.mode
    };
  }
}

class RoomManager {
  constructor() {
    this.rooms = new Map();
  }
  getOrCreateRoom(name) {
    if (!this.rooms.has(name)) {
      this.rooms.set(name, new RoomInfo(name));
    }
    return this.rooms.get(name);
  }
  getVisibleRooms() {
    return [...this.rooms.values()].filter(r => r.isVisible && r.isOpen);
  }
  prune() {
    for (const [k,r] of this.rooms) {
      if (r.playerCount === 0) this.rooms.delete(k);
    }
  }
  deleteRoom(name) {
    this.rooms.delete(name);
  }
}

class MasterServer {
  // gameServerAddresses: string[] — each is "ip:port", one per network interface.
  // When redirecting a client we pick the entry whose IP is closest (longest prefix
  // match) to the client's source address, so WSA clients get the WSA IP and
  // ZeroTier clients get the ZeroTier IP.
  constructor(bindAddress, port, gameServerAddresses, rooms) {
    this.bindAddress = bindAddress;
    this.port = port;
    // Accept both a single string and an array for backwards-compat
    this.gameServerAddresses = Array.isArray(gameServerAddresses)
      ? gameServerAddresses
      : [gameServerAddresses];
    this.rooms = rooms;
    this.sessions = new Set();
  }

  // Pick the advertised game-server address that best matches the client's IP.
  // "Best match" = most leading octets in common (longest prefix).
  pickGameServer(clientRemoteAddr) {
    if (this.gameServerAddresses.length === 1) return this.gameServerAddresses[0];
    // strip IPv6-mapped prefix (::ffff:) if present
    const clientIp = (clientRemoteAddr || '').replace(/^::ffff:/, '');
    const clientOctets = clientIp.split('.').map(Number);
    let best = this.gameServerAddresses[0];
    let bestScore = -1;
    for (const addr of this.gameServerAddresses) {
      const ip = addr.split(':')[0];  // addr is "ip:port"
      const octets = ip.split('.').map(Number);
      let score = 0;
      for (let i = 0; i < Math.min(octets.length, clientOctets.length); i++) {
        if (octets[i] === clientOctets[i]) score++;
        else break;
      }
      if (score > bestScore) { bestScore = score; best = addr; }
    }
    return best;
  }

  start() {
    this.server = net.createServer(sock => this.handleConnect(sock));
    this.server.listen(this.port, this.bindAddress, () => {
      console.log(`[Master] Listening on ${this.bindAddress}:${this.port}`);
      console.log(`[Master] Advertised game-server addresses: ${this.gameServerAddresses.join(', ')}`);
    });
    // periodic broadcast
    this.interval = setInterval(() => this.broadcastRoomList(false), 5000);
  }
  stop() {
    clearInterval(this.interval);
    this.server.close();
  }

  handleConnect(sock) {
    const session = new PlayerSession(sock);
    this.sessions.add(session);
    console.log(`[Master] Client ${session.id} connected from ${sock.remoteAddress}`);
    sock.on('data', data => this.handleData(session, data));
    sock.on('close', () => {
      console.log(`[Master] Client ${session.id} (${session.userId || 'anon'}) disconnected`);
      this.sessions.delete(session);
    });
  }

  handleData(session, data) {
    for (const frame of session.appendAndExtract(data)) {
      const msg = parseFrame(frame);
      if (!msg) continue;
      if (msg.IsInit) {
        session.send(Buffer.from([0xF3,0,0,0,9,0,0,0xF3,0x01]));
        continue;
      }
      if (msg.IsPing) {
        session.send(buildPingResponse(msg.PingFrame));
        continue;
      }
      if (msg.OpRequest) {
        this.processOp(session, msg.OpRequest);
      }
    }
  }

  processOp(session, req) {
    switch (req.OpCode) {
      case OpCode.Authenticate:
        this.handleAuth(session, req);
        break;
      case OpCode.JoinLobby:
        this.handleJoinLobby(session);
        break;
      case OpCode.CreateGame:
        this.handleCreateGame(session, req);
        break;
      case OpCode.JoinGame:
        this.handleJoinGame(session, req);
        break;
      case OpCode.JoinRandom:
        this.handleJoinRandom(session);
        break;
      default:
        console.log(`[Master] unhandled op 0x${req.OpCode.toString(16)}`);
        break;
    }
  }

  handleAuth(session, req) {
    if (req.Parameters && req.Parameters[ParamKey.userId]) {
      session.userId = String(req.Parameters[ParamKey.userId]);
    }
    console.log(`[Master] Session ${session.id} authenticated as '${session.userId}'`);
    session.isAuthenticated = true;
    session.send(buildOpResponse(OpCode.Authenticate, 0, null));
  }

  handleJoinLobby(session) {
    session.send(buildOpResponse(OpCode.JoinLobby, 0, null));
    this.sendRoomList(session, true);
  }

  handleCreateGame(session, req) {
    const roomName = (req.Parameters && req.Parameters[ParamKey.RoomName]) || Math.random().toString(36).slice(2,10);
    const room = this.rooms.getOrCreateRoom(roomName);
    console.log(`[Master] Client ${session.id} creating room '${roomName}'`);
    const gameAddr = this.pickGameServer(session.socket.remoteAddress);
    const p = { [ParamKey.GameServer]: gameAddr };
    session.send(buildOpResponse(OpCode.CreateGame, 0, p));
  }

  handleJoinGame(session, req) {
    const roomName = req.Parameters && req.Parameters[ParamKey.RoomName];
    if (!roomName) {
      session.send(buildOpResponse(OpCode.JoinGame, 32765, null, 'Room name missing'));
      return;
    }
    console.log(`[Master] Client ${session.id} joining room '${roomName}'`);
    const gameAddr = this.pickGameServer(session.socket.remoteAddress);
    const p = { [ParamKey.GameServer]: gameAddr };
    session.send(buildOpResponse(OpCode.JoinGame, 0, p));
  }

  handleJoinRandom(session) {
    const visible = this.rooms.getVisibleRooms();
    if (visible.length === 0) {
      session.send(buildOpResponse(OpCode.JoinRandom, 32760, null, 'NoRandomMatchFound'));
      return;
    }
    const room = visible[0];
    const gameAddr = this.pickGameServer(session.socket.remoteAddress);
    const p = { [ParamKey.RoomName]: room.name, [ParamKey.GameServer]: gameAddr };
    session.send(buildOpResponse(OpCode.JoinRandom, 0, p));
  }

  sendRoomList(session, full) {
    const eventCode = full ? 230 : 229; // GameList / GameListUpd
    const roomsHt = {};
    for (const r of this.rooms.getVisibleRooms()) {
      roomsHt[r.name] = {
        255: photonByte(r.maxPlayers),
        253: r.isOpen,
        254: r.isVisible,
        252: photonByte(r.playerCount),
        251: false,
        map:     r.map,
        version: r.version,
        mode:    r.mode
      };
    }
    const p = { 222: roomsHt };
    session.send(buildEvent(eventCode, p));
  }

  broadcastRoomList(full) {
    for (const s of this.sessions) {
      if (s.isAuthenticated) this.sendRoomList(s, full);
    }
  }
}

module.exports = { MasterServer, RoomManager };

File: LAN_Server_OptionB/src/MasterServer.cs
using System.Collections;
using System.Net;
using System.Net.Sockets;

namespace CNRLanServer;

/// <summary>
/// Listens on port 5055 (Photon master port).
/// Handles: Authenticate → JoinLobby → GameList event → CreateGame / JoinGame redirect.
///
/// After auth the client receives a game-list event immediately.
/// CreateGame and JoinGame both return the game-server address (port 5056).
/// The client then disconnects from master and connects to the game server.
/// </summary>
public class MasterServer
{
    private readonly int           _port;
    private readonly string        _gameServerAddress; // e.g. "192.168.1.10:5056"
    private readonly RoomManager   _rooms;
    private readonly CancellationTokenSource _cts = new();

    private readonly List<PlayerSession> _sessions = new();
    private readonly object              _sessLock  = new();

    public MasterServer(string bindAddress, int port, string gameServerAddress, RoomManager rooms)
    {
        _port              = port;
        _gameServerAddress = gameServerAddress;
        _rooms             = rooms;
    }

    public async Task RunAsync()
    {
        var listener = new TcpListener(IPAddress.Any, _port);
        listener.Start();
        Console.WriteLine($"[Master] Listening on port {_port}  →  redirecting to game server at {_gameServerAddress}");

        // Periodic room-list broadcast and prune
        _ = Task.Run(() => PeriodicBroadcastLoopAsync(_cts.Token));

        while (!_cts.IsCancellationRequested)
        {
            try
            {
                TcpClient client = await listener.AcceptTcpClientAsync(_cts.Token);
                var session = new PlayerSession(client);
                lock (_sessLock) _sessions.Add(session);
                _ = Task.Run(() => HandleClientAsync(session, _cts.Token));
            }
            catch (OperationCanceledException) { break; }
            catch (Exception ex)
            {
                Console.WriteLine($"[Master] Accept error: {ex.Message}");
            }
        }
        listener.Stop();
    }

    public void Stop() => _cts.Cancel();

    // ─── Client connection loop ───────────────────────────────────────────

    private async Task HandleClientAsync(PlayerSession session, CancellationToken ct)
    {
        Console.WriteLine($"[Master] Client {session.Id} connected from {((IPEndPoint?)session.Client.Client.RemoteEndPoint)?.Address}");
        try
        {
            var buffer = new byte[4096];
            while (session.IsAlive && !ct.IsCancellationRequested)
            {
                int read = await session.Stream.ReadAsync(buffer, 0, buffer.Length, ct);
                if (read == 0) break;

                Console.WriteLine($"[Master] Client {session.Id} raw recv ({read}b): {BitConverter.ToString(buffer, 0, Math.Min(read, 48))}");

                foreach (byte[] frame in session.AppendAndExtract(buffer, read))
                {
                    Console.WriteLine($"[Master] Client {session.Id} frame ({frame.Length}b): {BitConverter.ToString(frame, 0, Math.Min(frame.Length, 32))}");
                    await ProcessFrameAsync(session, frame, ct);
                }
            }
        }
        catch (OperationCanceledException) { }
        catch (Exception ex)
        {
            Console.WriteLine($"[Master] Client {session.Id} error: {ex.Message}");
        }
        finally
        {
            Console.WriteLine($"[Master] Client {session.Id} disconnected");
            lock (_sessLock) _sessions.Remove(session);
            session.Dispose();
        }
    }

    private async Task ProcessFrameAsync(PlayerSession session, byte[] frame, CancellationToken ct)
    {
        IncomingMessage? msg = PhotonBinaryProtocol.ParseFrame(frame);
        if (msg == null)
        {
            Console.WriteLine($"[Master] Client {session.Id} ParseFrame returned null for frame: {BitConverter.ToString(frame, 0, Math.Min(frame.Length, 32))}");
            return;
        }

        if (msg.IsInit)
        {
            Console.WriteLine($"[Master] Client {session.Id} init — AppId: '{msg.AppId}'");
            Console.WriteLine($"[Master] Sending InitResponse: {BitConverter.ToString(PhotonBinaryProtocol.InitResponse)}");
            await session.SendAsync(PhotonBinaryProtocol.InitResponse, ct);
            return;
        }

        if (msg.IsPing)
        {
            // Reply with proper 9-byte ping response (TConnect.Run always reads 9-byte header)
            await session.SendAsync(PhotonBinaryProtocol.BuildPingResponse(msg.PingFrame), ct);
            return;
        }

        if (msg.OpRequest == null) return;

        switch (msg.OpRequest.OpCode)
        {
            case OpCode.Authenticate:
                await HandleAuthAsync(session, msg.OpRequest, ct);
                break;

            case OpCode.JoinLobby:
                await HandleJoinLobbyAsync(session, ct);
                break;

            case OpCode.LeaveLobby:
                // No-op from our side; client just stops getting room updates
                await SendOkResponseAsync(session, OpCode.LeaveLobby, ct);
                break;

            case OpCode.CreateGame:
                await HandleMasterCreateGameAsync(session, msg.OpRequest, ct);
                break;

            case OpCode.JoinGame:
                await HandleMasterJoinGameAsync(session, msg.OpRequest, ct);
                break;

            case OpCode.JoinRandom:
                await HandleMasterJoinRandomAsync(session, ct);
                break;

            default:
                Console.WriteLine($"[Master] Client {session.Id} unhandled op: 0x{msg.OpRequest.OpCode:X2}");
                break;
        }
    }

    // ─── Op handlers ─────────────────────────────────────────────────────

    private async Task HandleAuthAsync(PlayerSession session, OpRequestData req, CancellationToken ct)
    {
        // Extract userId / app version if present (informational only)
        if (req.Parameters.TryGetValue(ParamKey.userId, out var uid))
            session.UserId = uid?.ToString() ?? "";
        if (req.Parameters.TryGetValue(225, out var nick)) // UserId sometimes carries nickname
            session.Nickname = nick?.ToString() ?? "";

        Console.WriteLine($"[Master] Client {session.Id} authenticated as '{session.UserId}'");

        // ReturnCode 0 = success; no required params needed for auth on master
        var resp = PhotonBinaryProtocol.BuildOpResponse(OpCode.Authenticate, 0, null);
        await session.SendAsync(resp, ct);
        session.IsAuthenticated = true;
    }

    private async Task HandleJoinLobbyAsync(PlayerSession session, CancellationToken ct)
    {
        await SendOkResponseAsync(session, OpCode.JoinLobby, ct);

        // Immediately push the current room list
        await SendRoomListEventAsync(session, ct, fullList: true);
    }

    private async Task HandleMasterCreateGameAsync(PlayerSession session, OpRequestData req, CancellationToken ct)
    {
        string roomName = req.Parameters.TryGetValue(ParamKey.RoomName, out var rn)
            ? (rn?.ToString() ?? Guid.NewGuid().ToString("N")[..8])
            : Guid.NewGuid().ToString("N")[..8];

        // Pre-create the room so it's visible via list before the client reaches the game server
        RoomInfo room = _rooms.GetOrCreateRoom(roomName);
        ApplyRoomPropsFromRequest(room, req);

        Console.WriteLine($"[Master] Client {session.Id} creating room '{roomName}' → redirecting to game server");

        var p = new Dictionary<byte, object>
        {
            [ParamKey.GameServer] = _gameServerAddress     // game server "ip:port"
        };
        await session.SendAsync(PhotonBinaryProtocol.BuildOpResponse(OpCode.CreateGame, 0, p), ct);
    }

    private async Task HandleMasterJoinGameAsync(PlayerSession session, OpRequestData req, CancellationToken ct)
    {
        string? roomName = req.Parameters.TryGetValue(ParamKey.RoomName, out var rn) ? rn?.ToString() : null;

        if (string.IsNullOrEmpty(roomName))
        {
            await session.SendAsync(
                PhotonBinaryProtocol.BuildOpResponse(OpCode.JoinGame, 32765, null, "Room name missing"), ct);
            return;
        }

        // We don't need the room to exist on master already; the game server is authoritative
        Console.WriteLine($"[Master] Client {session.Id} joining room '{roomName}' → redirecting to game server");

        var p = new Dictionary<byte, object>
        {
            [ParamKey.GameServer] = _gameServerAddress
        };
        await session.SendAsync(PhotonBinaryProtocol.BuildOpResponse(OpCode.JoinGame, 0, p), ct);
    }

    private async Task HandleMasterJoinRandomAsync(PlayerSession session, CancellationToken ct)
    {
        var rooms = _rooms.GetVisibleRooms();
        if (rooms.Count == 0)
        {
            // ErrorCode 32760 = NoRandomMatchFound
            await session.SendAsync(
                PhotonBinaryProtocol.BuildOpResponse(OpCode.JoinRandom, 32760, null, "NoRandomMatchFound"), ct);
            return;
        }

        var room = rooms[0]; // pick first available
        var p = new Dictionary<byte, object>
        {
            [ParamKey.RoomName]   = room.Name,
            [ParamKey.GameServer] = _gameServerAddress
        };
        await session.SendAsync(PhotonBinaryProtocol.BuildOpResponse(OpCode.JoinRandom, 0, p), ct);
    }

    // ─── Room list events ─────────────────────────────────────────────────

    public async Task SendRoomListEventAsync(PlayerSession session, CancellationToken ct, bool fullList = false)
    {
        if (!session.IsAuthenticated) return;

        byte eventCode = fullList ? EvtCode.GameList : EvtCode.GameListUpd;
        var roomsHt = BuildRoomListHashtable(fullList);

        var p = new Dictionary<byte, object>
        {
            [222] = roomsHt   // ParameterCode.GameList = 222
        };
        await session.SendAsync(PhotonBinaryProtocol.BuildEvent(eventCode, p), ct);
    }

    private Hashtable BuildRoomListHashtable(bool fullList)
    {
        var ht = new Hashtable();
        foreach (var room in _rooms.GetVisibleRooms())
        {
            // Per-room Hashtable: standard byte keys + string custom props
            var roomHt = new Hashtable
            {
                [(byte)255] = room.MaxPlayers,    // maxPlayers (byte)
                [(byte)253] = room.IsOpen,        // isOpen
                [(byte)254] = room.IsVisible,     // isVisible
                [(byte)252] = (byte)room.PlayerCount, // playerCount (byte)
                [(byte)251] = false,              // removedFromList = false
                // Custom string-keyed props
                ["map"]     = room.Map,
                ["version"] = room.Version,
                ["mode"]    = room.Mode
            };
            ht[room.Name] = roomHt;
        }
        return ht;
    }

    private async Task PeriodicBroadcastLoopAsync(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            await Task.Delay(TimeSpan.FromSeconds(5), ct).ContinueWith(_ => { });
            _rooms.Prune();

            List<PlayerSession> snapshot;
            lock (_sessLock) snapshot = new List<PlayerSession>(_sessions);

            foreach (var s in snapshot)
            {
                try { await SendRoomListEventAsync(s, ct, fullList: false); }
                catch { /* ignore disconnected sessions */ }
            }
        }
    }

    // ─── Helpers ──────────────────────────────────────────────────────────

    private static Task SendOkResponseAsync(PlayerSession session, byte opCode, CancellationToken ct) =>
        session.SendAsync(PhotonBinaryProtocol.BuildOpResponse(opCode, 0, null), ct);

    private static void ApplyRoomPropsFromRequest(RoomInfo room, OpRequestData req)
    {
        // Photon's OpCreateRoom encodes ALL room properties inside the GameProps Hashtable
        // (param 248), not at top-level params.  The structure sent by the client is:
        //
        //   req.Parameters[(byte)248] = Hashtable {
        //       (byte)255 = maxPlayers,
        //       (byte)253 = isOpen,
        //       (byte)254 = isVisible,
        //       (byte)250 = propsListedInLobby  (string[]),
        //       "version" = versionString,
        //       "map"     = mapName,
        //       "mode"    = modeString
        //   }
        //
        // Reading from top-level param (byte)239 / 237 / 236 finds nothing,
        // leaving maxPlayers=0 which causes the room-list filter (maxPlayers==playerCount)
        // to always hide the room.
        if (req.Parameters.TryGetValue(ParamKey.GameProps, out var gp) && gp is Hashtable gameProps)
        {
            if      (gameProps[byte.MaxValue] is byte   maxPb) room.MaxPlayers = maxPb;
            else if (gameProps[byte.MaxValue] is int    maxPi) room.MaxPlayers = (byte)maxPi;
            else if (gameProps[byte.MaxValue] is short  maxPs) room.MaxPlayers = (byte)maxPs;

            if (gameProps[(byte)253] is bool isOpenVal)  room.IsOpen    = isOpenVal;
            if (gameProps[(byte)254] is bool isVisVal)   room.IsVisible = isVisVal;

            if (gameProps["map"]     is string map)   room.Map     = map;
            if (gameProps["version"] is string ver)   room.Version = ver;
            if (gameProps["mode"]    is string mode)  room.Mode    = mode;
        }
    }
}
